src/server/conn/auto.rs

```diff
@@ -0,0 +1,409 @@
+    #[cfg_attr(docsrs, doc(cfg(feature = "http2")))]
+    pub fn http2_keep_alive_interval(
+        &mut self,
+        interval: impl Into<Option<Duration>>,
+    ) -> &mut Self {
+        self.http2.http2_keep_alive_interval(interval);
+        self
+    }
+    /// Sets a timeout for receiving an acknowledgement of the keep-alive ping.
+    ///
+    /// If the ping is not acknowledged within the timeout, the connection will
+    /// be closed. Does nothing if `http2_keep_alive_interval` is disabled.
+    ///
+    /// Default is 20 seconds.
+    ///
+    /// # Cargo Feature
+    ///
+    #[cfg(feature = "http2")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "http2")))]
+    pub fn http2_keep_alive_timeout(&mut self, timeout: Duration) -> &mut Self {
+        self.http2.http2_keep_alive_timeout(timeout);
+        self
+    }
+    /// Set the maximum write buffer size for each HTTP/2 stream.
+    ///
+    /// Default is currently ~400KB, but may change.
+    ///
+    /// # Panics
+    ///
+    /// The value must be no larger than `u32::MAX`.
+    #[cfg(feature = "http2")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "http2")))]
+    pub fn http2_max_send_buf_size(&mut self, max: usize) -> &mut Self {
+        self.http2.http2_max_send_buf_size(max);
+        self
+    }
+    /// Enables the [extended CONNECT protocol].
+    ///
+    /// [extended CONNECT protocol]: https://datatracker.ietf.org/doc/html/rfc8441#section-4
+    #[cfg(feature = "http2")]
+    pub fn http2_enable_connect_protocol(&mut self) -> &mut Self {
+        self.http2.http2_enable_connect_protocol();
+        self
+    }
+    /// Sets the max size of received header frames.
+    ///
+    /// Default is currently ~16MB, but may change.
+    #[cfg(feature = "http2")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "http2")))]
+    pub fn http2_max_header_list_size(&mut self, max: u32) -> &mut Self {
+        self.http2.http2_max_header_list_size(max);
+        self
+    }
+    /// Set the timer used in background tasks.
+    pub fn http2_timer<M>(&mut self, timer: M) -> &mut Self
+    where
+        M: Timer + Send + Sync + 'static,
+    {
+        self.http2.timer(timer);
+        self
+    }
+
+    /// Bind a connection together with a [`Service`].
+    pub async fn serve_connection<I, S, B>(&self, mut io: I, service: S) -> Result<()>
+    where
+        S: Service<Request<Incoming>, Response = Response<B>> + Send,
+        S::Future: Send + 'static,
+        S::Error: Into<Box<dyn StdError + Send + Sync>>,
+        B: Body + Send + 'static,
+        B::Data: Send,
+        B::Error: Into<Box<dyn StdError + Send + Sync>>,
+        I: AsyncRead + AsyncWrite + Unpin + 'static,
+    {
+        let mut buf = Vec::new();
+
+        let protocol = loop {
+            if buf.len() < 24 {
+                io.read_buf(&mut buf).await?;
+
+                let len = buf.len().min(H2_PREFACE.len());
+
+                if buf[0..len] != H2_PREFACE[0..len] {
+                    break Protocol::H1;
```

Comment:
The HTTP/1 stack has an error code for "this is an HTTP/2 stream" so it's possible to just start with HTTP/1 normally and then rewind and switch to HTTP/2 if you see that error.  Uncertain which is nicer/better but just mentioning it in case it's interesting.

Something like https://github.com/hyperium/hyper/pull/3084/files#diff-4e236d9f0f3f026b7b233382e56a937697d22b5be15b6faf38a479fc722de25aR550-R597 (which I wrote before I saw this MR)

Reply:
Current approach makes more sense IMO.

Reply:
Thinking about it - I agree (and, because this is a "serve_connection" call, it fits into my FFI usage just fine 👍)